### PR TITLE
impl(generator): add LRO support to rest stub and decorators

### PR DESCRIPTION
**Pull request by scotthart**
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10624)

### CODE REVIEW

File: generator/integration_tests/tests/golden_thing_admin_rest_stub_test.cc [hunk 1]
```diff
@@ -114,8 +118,12 @@ TEST(GoldenThingAdminRestStubTest, ListDatabases) {
   EXPECT_THAT(success->next_page_token(), Eq("my_next_page_token"));
 }
 
-TEST(GoldenThingAdminRestStubTest, CreateDatabase) {
-  auto mock_rest_client = absl::make_unique<MockRestClient>();
+TEST(GoldenThingAdminRestStubTest, AsyncCreateDatabase) {
+  auto mock_cq = std::make_shared<MockCompletionQueueImpl>();
+  EXPECT_CALL(*mock_cq, RunAsync)
```
Comment (coryan) [inline generator/integration_tests/tests/golden_thing_admin_rest_stub_test.cc:123]: If you want, you could delay this execution until later (maybe using `AsyncSequencer`) to verify your code handles slow thread starts correctly.

**Review by coryan — CHANGES_REQUESTED** [review]

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10624)* status: 0 of 17 files reviewed, 7 unresolved discussions (waiting on @coryan)

___
*[`generator/internal/logging_decorator_rest_generator.cc` line 81 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10624#-NMBRLHYdcqSlHMbcn5k-r1-81:-NMFkp_VBoW2fqLX6srC:b-wryfwj) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/a11d8692d3f39192dd224c45935e9fc4967bd7a4/generator/internal/logging_decorator_rest_generator.cc#L81)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

nit: I would have said `continue; }` instead of `} else {`.  Your call.
</blockquote></details>

Left as `else` for the time being.
___
*[`generator/internal/stub_rest_generator.cc` line 71 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10624#-NMBRLHYdcqSlHMbcn5m-r1-71:-NMFktyvAnFAWIdWwtBL:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/a11d8692d3f39192dd224c45935e9fc4967bd7a4/generator/internal/stub_rest_generator.cc#L71)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Maybe use the same style as in the other files to skip unhandled method types?
</blockquote></details>

Done.
___
*[`generator/internal/stub_rest_generator.cc` line 98 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10624#-NMBRLHYdcqSlHMbcn5m-r1-98:-NMFkvIc6AAN7u4Hbyy3:b-hd0d06) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/a11d8692d3f39192dd224c45935e9fc4967bd7a4/generator/internal/stub_rest_generator.cc#L98)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Do we need to skip methods without HTTP annotations here?
</blockquote></details>

Added skip.
___
*[`generator/internal/stub_rest_generator.cc` line 259 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10624#-NMBRLHYdcqSlHMbcn5m-r1-259:-NMFkxdcEyM8Qvd7iDjm:b-u1cmsy) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/a11d8692d3f39192dd224c45935e9fc4967bd7a4/generator/internal/stub_rest_generator.cc#L259)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Nit: seems like you could consolidate the `if (HasLongRunningMethod())` blocks?
</blockquote></details>

Consolidated.
___
*[`generator/internal/stub_rest_generator.cc` line 297 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10624#-NMBRLHYdcqSlHMbcn5m-r1-297:-NMFkzH-3c7DmcZO0Qsx:b-hgnlt0) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/a11d8692d3f39192dd224c45935e9fc4967bd7a4/generator/internal/stub_rest_generator.cc#L297)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

You are capturing the `p` by reference and then using it in another thread.  The thread may not start until this function returns.  That is a dangling reference.  I think you want:

```cc
  auto f = p.get_future();
  cq.RunAsync([p = std::move(p), this] { ...
```

Some other things are also a bit risky:

- `rest_context` you want to `std::move()` that too, but that suggests the API should consume a `std::unique_ptr<>`
- You probably need to copy `request`.
- I think `*service_` you may capture by reference, but maybe a copy is cleaner?

</blockquote></details>

Fixed everything except `RestContext` lifetime. I'll fix that in a separate PR.
___
*[`generator/internal/stub_rest_generator.cc` line 343 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10624#-NMBRLHYdcqSlHMbcn5m-r1-343:-NMFl98tDtFXjWUndYPm:bribihm) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/a11d8692d3f39192dd224c45935e9fc4967bd7a4/generator/internal/stub_rest_generator.cc#L343)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Similar comments as above w.r.t. to local object lifetimes.
</blockquote></details>

As above.

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10624)* status: 0 of 17 files reviewed, 7 unresolved discussions (waiting on @coryan)

___
*[`generator/integration_tests/tests/golden_thing_admin_rest_stub_test.cc` line 123 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10624#-NMFkld404aiEi2CUQ3b-r2-123:-NMFlMhj89CHpTFHw6rq:b-5g129) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/2f4b6d0a2cd9d290e785c256401b564c4d89fb58/generator/integration_tests/tests/golden_thing_admin_rest_stub_test.cc#L123)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

If you want, you could delay this execution until later (maybe using `AsyncSequencer`) to verify your code handles slow thread starts correctly.
</blockquote></details>

Switched to actual CQ implementation.

**Review by coryan — CHANGES_REQUESTED** [review]
> Reviewed 4 of 10 files at r3, 2 of 3 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10624)* status: 6 of 17 files reviewed, 3 unresolved discussions (waiting on @scotthart)

___
*[`generator/internal/stub_rest_generator.cc` line 297 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10624#-NMBRLHYdcqSlHMbcn5m-r1-297:-NMG8mDc3jSoYxcDyn4-:b-159zsj) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/a11d8692d3f39192dd224c45935e9fc4967bd7a4/generator/internal/stub_rest_generator.cc#L297)):*
<details><summary><i>Previously, scotthart (Scott Hart) wrote…</i></summary><blockquote>

Fixed everything except `RestContext` lifetime. I'll fix that in a separate PR.
</blockquote></details>

Hmmm... I think you might be introducing a flake.  The *san builds will fail (sometimes, if they run the golden tests) if the thread starts very slowly.  Other builds may simply crash from time to time.
___
*[`generator/internal/stub_rest_generator.cc` line 301 at r4](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10624#-NMG33FE8hSBO913KuSI:-NMG33FE8hSBO913KuSJ:b-14gii8) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/a598a1c73910c9204e50b0fb46e78c98aeb671b5/generator/internal/stub_rest_generator.cc#L301)):*
> ```cpp
>   promise<StatusOr<google::longrunning::Operation>> p;
>   future<StatusOr<google::longrunning::Operation>> f = p.get_future();
>   std::thread t1{[service = service_, request, &rest_context](auto p) {
> ```

Any reason to capture some parameters and pass others as arguments?  You could pass `service`, `request` and maybe `std::ref(rest_context)` as arguments too?
___
*[`generator/internal/stub_rest_generator.cc` line 306 at r4](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10624#-NMG3fW8BCwaG2aB3uH4:-NMG3fW8BCwaG2aB3uH5:b-er49n3) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/a598a1c73910c9204e50b0fb46e78c98aeb671b5/generator/internal/stub_rest_generator.cc#L306)):*
> ```cpp
>           $method_rest_path$$method_http_query_parameters$));
>   }, std::move(p)};
>   return f.then([t2 = std::move(t1), cq](auto f) mutable {
> ```

FYI,  if `[t = std::move(t), cq]` works maybe you can avoid the different names.  Your call.

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10624)* status: 6 of 17 files reviewed, 3 unresolved discussions (waiting on @coryan)

___
*[`generator/internal/stub_rest_generator.cc` line 297 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10624#-NMBRLHYdcqSlHMbcn5m-r1-297:-NMGIgSe9IufO3DOxbQd:bhi0p3) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/a11d8692d3f39192dd224c45935e9fc4967bd7a4/generator/internal/stub_rest_generator.cc#L297)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Hmmm... I think you might be introducing a flake.  The *san builds will fail (sometimes, if they run the golden tests) if the thread starts very slowly.  Other builds may simply crash from time to time.
</blockquote></details>

I'll disable these tests until the issue is fixed.
___
*[`generator/internal/stub_rest_generator.cc` line 301 at r4](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10624#-NMG33FE8hSBO913KuSI:-NMGJUyd6SybxLP6fXpo:bmzjdvx) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/a598a1c73910c9204e50b0fb46e78c98aeb671b5/generator/internal/stub_rest_generator.cc#L301)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Any reason to capture some parameters and pass others as arguments?  You could pass `service`, `request` and maybe `std::ref(rest_context)` as arguments too?
</blockquote></details>

Moved everything but the rest_context to parameters. We still need to capture the `RestContext&` until we can change it to a `std::unique_ptr<RestContext>` due to thread rvalue argument requirements.
___
*[`generator/internal/stub_rest_generator.cc` line 306 at r4](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10624#-NMG3fW8BCwaG2aB3uH4:-NMGIl_b2n_4zXe-Qckg:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/a598a1c73910c9204e50b0fb46e78c98aeb671b5/generator/internal/stub_rest_generator.cc#L306)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

FYI,  if `[t = std::move(t), cq]` works maybe you can avoid the different names.  Your call.
</blockquote></details>

Done.

**Review by coryan — APPROVED** [review]
:lgtm:

> Reviewed 1 of 3 files at r1, 3 of 4 files at r2, 5 of 10 files at r3, 3 of 3 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10624#-:-NMGL5akCcKfCXsOYs_Z:bnfp4nl)* status: 16 of 17 files reviewed, all discussions resolved

**Pull request merged**